### ISSUE TITLE
[Question] What are the configurations of the Cassandra database in the tb-cassandra image and how can they be changed in the Compose file?

### ISSUE THREAD

**Issue by BAslan1988** [thingsboard/thingsboard#8199]
**Component**

 * Installation

**Description**

While setting up a Thingsboard instance, I noticed that I cannot easily change the configurations of the predefined Cassandra database. After an Out.of.memory exception in the Cassandra logs, I wanted to allocate more memory to the application, but I can't find the corresponding environment variable. 

So my question how can I change configurations of predefined Cassandra Database of the tb-cassandra Image?

Another Question is, that I cant' have access to the folders although i set up volumes for them. I also checked Read/Write rights on the hostsystem's directories.

**Environment**

 * OS:  RockyLinux 5.14.0
 * ThingsBoard CE: 3.4.3
 * Browser: Chrome
 * CPU: Intel(R) Xeon(R) Gold 6146 CPU @ 3.20GHz
 * RAM: 12 GB

**Docker-Compose.yml**

version: '3.0'
services:
  zookeeper:
    container_name: zookeeper
    restart: always
    image: "zookeeper:3.5"
    ports:
      - "2181:2181"
    environment:
      ZOO_MY_ID: 1
      ZOO_SERVERS: server.1=zookeeper:2888:3888;zookeeper:2181
    volumes:
      - tb_zookeeper_data:/data
      - tb_zookeeper_logs:/logs
      - tb_zookeeper_datalogs:/datalog
  kafka:
    container_name: kafka
    restart: always
    image: wurstmeister/kafka
    depends_on:
      - zookeeper
    ports:
      - "9092:9092"
    environment:
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
      KAFKA_LISTENERS: INSIDE://:9093,OUTSIDE://:9092
      KAFKA_ADVERTISED_LISTENERS: INSIDE://:9093,OUTSIDE://kafka:9092
      KAFKA_LISTENER_SECURITY_PROTOCOL_MAP: INSIDE:PLAINTEXT,OUTSIDE:PLAINTEXT
      KAFKA_INTER_BROKER_LISTENER_NAME: INSIDE
    volumes:
      - kafka:/run/podman/podman.sock
      - kafka_:/kafka
  thingsboard_ce:
    container_name: thingsboard
    restart: always
    image: "thingsboard/tb-cassandra"
    depends_on:
      - kafka
    ports:
      - "8080:9090"
      - "1883:1883"
      - "7070:7070"
      - "5683-5688:5683-5688/udp"
    environment:
      TB_QUEUE_TYPE: kafka
      TB_KAFKA_SERVERS: kafka:9092
      JAVA_OPTS: "-Xms2g -Xmx8g"
    volumes:
      - tb_logs:/usr/share/thingsboard/logs/
      - tb_data:/usr/share/thingsboard/data/
      - tb_conf:/usr/share/thingsboard/conf
      - tb_cassandra_data:/data
     logging:
       driver: gelf
       gelf-address: udp://xxx.xx.xxx.xx:xxxx #Graylog Adress
       tag: "Thingsboard/{{.Name}}"

**Comment by devaskim** [comment]
Maybe [Dockerfile](https://github.com/thingsboard/thingsboard/blob/master/msa/tb/docker-cassandra/Dockerfile) of the image helps you.

You could mount configuration file on host machine to the instance of the image.

**Comment by BAslan1988** [comment]
Thank you for providing the Dockerfile. I will have a look on it.

**After another slow perfomance of Thingsboard's visualization of  Timesseries Charts I looked up the cassandra logs (system.log) and found  an OOM Exception:**

ERROR [GossipTasks:1] 2023-03-12 03:30:42,748 DefaultPromise.java:844 - Failed to submit a listener notification task. Event loop shut down?
java.lang.OutOfMemoryError: unable to create native thread: possibly out of memory or process/resource limits reached
	at java.base/java.lang.Thread.start0(Native Method)
	at java.base/java.lang.Thread.start(Thread.java:798)
	at io.netty.util.concurrent.GlobalEventExecutor.startThread(GlobalEventExecutor.java:232)
	at io.netty.util.concurrent.GlobalEventExecutor.execute(GlobalEventExecutor.java:208)
	at io.netty.util.concurrent.DefaultPromise.safeExecute(DefaultPromise.java:842)
	at io.netty.util.concurrent.DefaultPromise.notifyListenerWithStackOverFlowProtection(DefaultPromise.java:529)
	at io.netty.util.concurrent.DefaultPromise.notifyListener(DefaultPromise.java:477)
	at io.netty.util.concurrent.CompleteFuture.addListener(CompleteFuture.java:48)
	at org.apache.cassandra.net.FutureCombiner.<init>(FutureCombiner.java:83)
	at org.apache.cassandra.net.FutureCombiner.<init>(FutureCombiner.java:55)
	at org.apache.cassandra.net.InboundSockets.open(InboundSockets.java:235)
	at org.apache.cassandra.net.MessagingService.waitUntilListening(MessagingService.java:534)
	at org.apache.cassandra.gms.Gossiper$GossipTask.run(Gossiper.java:287)
	at org.apache.cassandra.concurrent.DebuggableScheduledThreadPoolExecutor$UncomplainingRunnable.run(DebuggableScheduledThreadPoolExecutor.java:118)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)



**My Question is the OOM Problem related to a wrong memory Allocation of the JVM or are there just too many threads active, which the CPU (1 Core CPU) can't handle? Do I have to change concurrent write and read Operation from default 32 to another number? Thanks in advance.**

**Comment by BAslan1988** [comment]
Update: I configured cassandra configuration according to  [Amy's Cassandra tuning guide](https://tobert.github.io/pages/als-cassandra-21-tuning-guide.html).

The display of the time series charts is still slow in the browser. It get worse as soon as several clients access the dashboard via the browser, there are delays in the display. 

Do you had similiar observations and is there any solution for this problem?

**Another Question is**:  I looked up the thingsboard.yml and saw, that the configuration of the database usage for timeseries is **SQL** althoug  used the tb-cassandra image:

![image](https://user-images.githubusercontent.com/116588892/226087909-eaf5e14f-01ab-4442-802e-a0f2a57b94a1.png)

Is this intended?

**Comment by devaskim** [comment]
Cassandra option is [here](https://github.com/thingsboard/thingsboard/blob/master/msa/tb/docker-cassandra/Dockerfile#L24), so Cassandra TB image uses Cassandra database.
I haven't no exact anwer for performance question, it depends on many things. But here is [Troubleshooting](https://thingsboard.io/docs/user-guide/troubleshooting/) and [Performance ](https://thingsboard.io/docs/reference/performance/) pages. Maybe it helps you.

**Comment by BAslan1988** [comment]
Hello, we have tested the system so far and have come to the conclusion, that the bottleneck here seems to be the fact, that only a single cpu core is managing and processing the the whole application.

For further investigations we enabled thingsboard-prometheus metrics to observe the performance.

 ```METRICS_ENABLED = true and METRICS_ENDPOINTS_EXPOSE = prometheus```

Thingsboard shows some interesting metrics to track according to your suggested Troubleshooting-Link but there is no complete list of all metrics provided for prometheus. 

For example I can scrape JVM metrics but I dont know to which component these metrics are refered to.
Is it Cassandra ( Datastax-driver), Postgresql or Thingsboard itself?

My Question is:
Is there any description of every metric provided from thingsboard via prometheus?

Thanks in advance
